docs(ai-agents): Add AI connector tutorials and Agent engine positioning (#70461)
airbytehq/airbyte: docs/ai-agents/connectors/tutorial-python.md

---
sidebar_label: "Python SDK tutorial"
---

# Get started with AI connectors: Python SDK

In this tutorial, you'll create a new Python project with `uv`, add a Pydantic AI agent, equip it to use one of Airbyte's AI connectors, and use natural language to explore your data. This tutorial uses GitHub, but if you don't have a GitHub account, you can use one of Airbyte's other AI connectors and perform different operations.

> Using the Python SDK is more time-consuming than the Connector MCP server, but affords you the most control over how you use AI connectors.

Reviewer: ```suggestion
Using the Python SDK is more time-consuming than the Connector MCP server, but affords you the most control over the context you send to your agent.
```

- [uv](https://github.com/astral-sh/uv)
- A [GitHub personal access token](https://github.com/settings/tokens). For this tutorial, a classic token with `repo` scope is sufficient.
- An [OpenAI API key](https://platform.openai.com/api-keys). This tutorial uses OpenAI, but Pydantic AI supports other LLM providers if you prefer.

## Part 1: Create a new Python project

In this tutorial you initialize a basic Python project to work in. However, if you have an existing project you want to work with, feel free to use that instead.

1. Create a new project using uv:

   ```bash
   uv init my-ai-agent --app
   cd my-ai-agent
   ```

   This creates a project with the following structure:

   ```text
   my-ai-agent/
   ├── .gitignore
   ├── .python-version
   ├── README.md
   ├── main.py
   └── pyproject.toml
   ```

2. Create an `agent.py` file for your agent definition:

   ```bash
   touch agent.py
   ```

You create `.env` and `uv.lock` files in later steps, so don't worry about them yet.

## Part 2: Install dependencies

Install the GitHub connector and Pydantic AI. This tutorial uses OpenAI as the LLM provider, but Pydantic AI supports many other providers.

```bash

> uv add airbyte-ai-github pydantic-ai

Reviewer: note to @girarda we need to rename this

```

This command installs:

> - `airbyte-ai-github`: The Airbyte AI connector for GitHub, which provides type-safe access to GitHub's API.

Reviewer: @girarda reminder to rename this

### Define the connector

Define the AI connector for GitHub. It authenticates using your personal access token.

```python title="agent.py"
connector = GithubConnector(
    auth_config=GithubAuthConfig(
        access_token=os.environ["GITHUB_ACCESS_TOKEN"]
    )
)
```

### Create the agent

Create a Pydantic AI agent with a system prompt that describes its purpose:

```python title="agent.py"
agent = Agent(
    "openai:gpt-4o",
    system_prompt=(
        "You are a helpful assistant that can access GitHub repositories, issues, "
        "and pull requests. Use the available tools to answer questions about "
        "GitHub data. Be concise and accurate in your responses."
    ),
)
```

- The `"openai:gpt-4o"` string specifies the model to use. You can use a different model by changing the model string. For example, use `"openai:gpt-4o-mini"` to lower costs, or see the [Pydantic AI models documentation](https://ai.pydantic.dev/models/) for other providers like Anthropic or Google.
- The `system_prompt` parameter tells the LLM what role it should play and how to behave.

## Part 6: Add tools to your agent

> Tools let your agent fetch real data from GitHub using Airbyte's AI connector. Without tools, the agent can only respond based on its training data. By registering connector operations as tools, the agent can decide when to call them based on natural language questions.

Reviewer: ```suggestion
Tools let your agent fetch real data from GitHub using Airbyte's Agent connector. Without tools, the agent can only respond based on its training data. By registering connector operations as tools, the agent can decide when to call them based on natural language questions.
```
